pkg/resourcemanager/azuresql/azuresqlfailovergroup/azuresqlfailovergroup_reconcile.go

```diff
@@ -35,84 +37,102 @@ func (fg *AzureSqlFailoverGroupManager) Ensure(ctx context.Context, obj runtime.
 	}
 
-	groupName := instance.Spec.ResourceGroup
-	serverName := instance.Spec.Server
+	pClient := pollclient.NewPollClient(fg.Creds)
+	lroPollResult, err := pClient.PollLongRunningOperationIfNeeded(ctx, &instance.Status)
+	if err != nil {
+		instance.Status.Message = err.Error()
+		return false, err
+	}
+	if lroPollResult == pollclient.PollResultTryAgainLater {
+		// Need to wait a bit before trying again
+		return false, nil
+	}
+
+	failoverGroupsClient, err := azuresqlshared.GetGoFailoverGroupsClient(fg.Creds)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to create failovergroup client")
+	}
+
+	notFoundErrorCodes := []string{
+		errhelp.ParentNotFoundErrorCode,
+		errhelp.ResourceGroupNotFoundErrorCode,
+		errhelp.NotFoundErrorCode,
+		errhelp.ResourceNotFound,
+	}
+
 	failoverGroupName := instance.ObjectMeta.Name
-	sqlFailoverGroupProperties := azuresqlshared.SQLFailoverGroupProperties{
-		FailoverPolicy:               v1alpha1.ReadWriteEndpointFailoverPolicy(instance.Spec.FailoverPolicy),
-		FailoverGracePeriod:          instance.Spec.FailoverGracePeriod,
-		SecondaryServer:              instance.Spec.SecondaryServer,
-		SecondaryServerResourceGroup: instance.Spec.SecondaryServerResourceGroup,
-		DatabaseList:                 instance.Spec.DatabaseList,
+	azureFailoverGroup, err := failoverGroupsClient.Get(ctx, instance.Spec.ResourceGroup, instance.Spec.Server, failoverGroupName)
+	if err != nil {
+		azerr := errhelp.NewAzureError(err)
+		if azerr.Type != errhelp.ResourceNotFound {
+			instance.Status.Message = fmt.Sprintf("AzureSqlFailoverGroup Get error %s", err.Error())
+			return errhelp.HandleEnsureError(err, notFoundErrorCodes, nil)
+		}
 	}
 
-	resp, err := fg.GetFailoverGroup(ctx, groupName, serverName, failoverGroupName)
-	if err == nil {
+	failoverGroupProperties, err := fg.TransformToSQLFailoverGroup(ctx, instance)
+	if err != nil {
+		instance.Status.Message = err.Error()
+		return errhelp.HandleEnsureError(err, notFoundErrorCodes, nil)
+	}
+
+	// We found a failover group, check to make sure that it matches what we have locally
+	doesResourceMatchAzure := DoesResourceMatchAzure(failoverGroupProperties, azureFailoverGroup)
 
-		if *resp.ReplicationState == "SEEDING" {
+	if doesResourceMatchAzure {
+		if *azureFailoverGroup.ReplicationState == "SEEDING" {
+			instance.Status.Message = "FailoverGroup replicationState is SEEDING"
 			return false, nil
 		}
 
-		instance.Status.Provisioning = false
-		instance.Status.Provisioned = true
-		instance.Status.Message = resourcemanager.SuccessMsg
-		instance.Status.ResourceId = *resp.ID
+		key := types.NamespacedName{Name: instance.ObjectMeta.Name, Namespace: instance.Namespace}
+		_, err := fg.SecretClient.Get(ctx, key)
+		// We make the same assumption many other places in the code make which is that if we cannot
+		// get the secret it must not exist.
+		if err != nil {
+			// We may need to create a secret
+			secret := fg.GetOrPrepareSecret(ctx, instance)
```

Comment:
`GetOrPrepareSecret` tries to get the secret internally too - maybe change this to just create a new secret for upserting?

Reply:
Good catch - fixed